skip motd in UI if request initiated from tsh headless auth

flyinghermit opened the pull request:
When login in Web UI is initiated for headless auth, for example with authentication link generated by `tsh --headless` command: 
```
Complete headless authentication in your local web browser:

https://<proxy>/web/headless/c7bd5f24-d487-5dd4-b30e-371a80bd5993
```
opening the link in browser redirects to login page with the following redirect format:
`https://<proxy>/web/login?redirect_uri=https://<proxy>/web/headless/c7bd5f24-d487-5dd4-b30e-371a80bd5993`

To disable MOTD when auth is initiated from headless auth, now we set `showMotd = false` if login url includes `redirect_uri` and the `redirect_uri` includes `headless`.

Tested with both local user and Github user configured with Github auth.

closes https://github.com/gravitational/teleport/issues/28746

flyinghermit committed:
skip motd for headless auth

sshahcodes commented:
FYI @zmb3, I have manually tested this and it seems to do the trick. Tested with both local user account and for user connected with Github auth. LMK if there's other scenario I may have missed.

zmb3 commented on web/packages/teleport/src/Login/useLogin.ts:
```diff
@@ -39,7 +39,16 @@ export default function useLogin() {
-  const [showMotd, setShowMotd] = useState(!!motd);
+  const [showMotd, setShowMotd] = useState(() => {
+    const url = new URL(window.location.href);
+    const redirect_uri = url.searchParams.get('redirect_uri');
```
```suggestion
    const redirectUri = url.searchParams.get('redirect_uri');
```

zmb3 commented on web/packages/teleport/src/Login/useLogin.ts:
```diff
@@ -39,7 +39,16 @@ export default function useLogin() {
+      return false;
+    } else if (cfg.getMotd()) {
+      return true;
+    }
```
This is not exhaustive right? I prefer we return a Boolean in all cases and don't leave undefined as a possibility.

gzdunek commented on web/packages/teleport/src/Login/useLogin.ts:
```diff
@@ -39,7 +39,16 @@ export default function useLogin() {
   const motd = cfg.getMotd();
-  const [showMotd, setShowMotd] = useState(!!motd);
+  const [showMotd, setShowMotd] = useState(() => {
+    const url = new URL(window.location.href);
```
Nit: I think it is better to use `react-router` APIs, instead of accessing and operating on `window` directly. 
```tsx
  const location = useLocation();
  const [showMotd, setShowMotd] = useState(() => {
    const searchParams = new URLSearchParams(location.search);
    const redirectUri = searchParams.get('redirect_uri');
...
```

sshahcodes replied:
Following your suggestion, I changed the code to use `useLocation`. But then found it actually adds a dependency to all the tests used in `Login.test.tsx: Jest: TypeError: Cannot read properties of undefined (reading 'location')`. That seems to be solved by wrapping the test component to render inside react router (e.g. `<MemoryRouter><Login /><MemoryRouter>`). But the fact that I will have to update other tests which depends on location: https://github.com/gravitational/teleport/blob/884bd7fb3847c8173c50a3f28cbfe1fa7e625e66/web/packages/teleport/src/Login/Login.test.tsx#L27-L31 I am cautious if this change would add any extra value given the risk of bugging existing test logic (tests are passing locally after wrapping inside `<MemoryRouter />` but still...)

What do you think? I recon it would be following recommended practice to use `react-router` but wanted to check on your opinion here.

gzdunek replied:
Ah, probably it doesn't make sense to add another dependency there. 
However, I think we could use `history.getLocation()`, at lest we will relay on a single API.

But if this won't work, I'm fine with `window.location`

sshahcodes replied:
`history.getLocation()` had the same effect, but `history.getRedirectParam()` worked like charm. Used that instead of direct `window.location` in https://github.com/gravitational/teleport/pull/31050/commits/73b6f2e93082206a12a0b90415c3d11719ef0551

Thanks for pointing to that.

flyinghermit committed:
address review suggestion:
- rename redirect_uri to redirectUri
- add explicit default showMotd type and setShowMotd return type

flyinghermit committed:
group motd tests

flyinghermit committed:
use history.getRedirectParam()

gzdunek commented on web/packages/teleport/src/Login/Login.test.tsx:
```diff
@@ -114,41 +115,65 @@ test('login with private key policy enabled through role setting', async () => {
+    const url =
+      'https://teleport.example.com/web/login?redirect_uri=https://teleport.example.com/web/headless/5c5c1f73-ac5c-52ee-bc9e-0353094dcb4a';
+    render(
+      <MemoryRouter>
```
Unfortunately, this test doesn't really work. `Route` component doesn't set the active route to the given path, but instead it tells the `Router`: "render my children when the active location equals my `path`".
So, this test passes because nothing is rendered :)
What's more, in `beforeEach` we mock the `history.getRedirectParam` so it always returns `/`.  

IMO, the best way to test this it to mock this function in our test:
```ts
jest.spyOn(history, 'getRedirectParam').mockReturnValue('https://teleport.example.com/web/headless/5c5c1f73-ac5c-52ee-bc9e-0353094dcb4a');
```
and remove `MemoryRouter` and `Route`.

gzdunek commented on web/packages/teleport/src/Login/useLogin.ts:
```diff
@@ -39,7 +39,17 @@ export default function useLogin() {
+  const [showMotd, setShowMotd] = useState<boolean>(() => {
+    const redirectUri = history.getRedirectParam();
+
+    if (redirectUri?.includes('headless')) {
+      return false;
+    } else if (cfg.getMotd()) {
+      return true;
+    } else {
+      return false;
+    }
```
```suggestion
    if (redirectUri?.includes('headless')) {
      return false;
    }
    return !!cfg.getMotd();
```

sshahcodes replied:
this is much better. updated in https://github.com/gravitational/teleport/pull/31050/commits/4edc19f866d066fed7268b5d586f7c8e56592634

flyinghermit committed:
fix test: mock getRedirectParam, remove nested motd check
